contributors/design-proposals/storage/local-storage-pv.md

```diff
@@ -761,3 +761,124 @@ The central manager will poll each worker for their discovered volumes and creat
+that is:
+
+1. When feature `DynamicProvisioningScheduling` enabled, scheduler would verify a pod's volume requirement,
+and set `annScheduledNode` on PVCs that need provisioning.
+2. New fields would be added in `StorageClass` to allow storage providers to report
+capacity limits per topology domain.
+
+The local volume provisioner need to be updated in several aspects to support the scenario.
+
+##### Backend storage management
+
+Mechanisms to manage local storage sources are varied. They could be LVM, Ceph, or something like that.
+In addition to current `populator`, `deletor` and `discoverer`, a new component named
+`storageManager` will be introduced in the plugin, to manage the backend storage.
+
+`storageManager` is a general interface, each kind of storage should have its own implementation.
+During start-up, a new environment variable `STORAGE_BACKEND` is set in the provisioner DaemonSet,
+to determine which StorageManager instance to start.
+(TBD: Should we support the scenario of multi backends on a node?)
+
+A StorageManager instance is responsible for:
+
+1. Monit changes in storage input, and handle corresponding initialization.
+2. Update capacity in StorageClass when storage source changed.
+3. Provide the ability to create volumes, detailed in the volume creation section.
+
+Take LVM as an example, when a new source disk "/dev/new" is added to its StorageClass `exStorage`,
+the StorageManager should:
+
+1. Check if related PV (Physical volume in LVM) exists, if not, create one with command:
+```
+pvcreate /dev/new
+```
+
+2. Check if the Physical volume exists in Volume group exStorageGroup, if not, add it:
+```
+vgextend exStorageGroup /dev/new
+```
+
+3. Update capacity in StorageClass if there is a change in VG size.
+
+A new field `storageSource` will be added into the configuration ConfigMap, to illustrate available
+storage source for provisioning.
+
+```
+kind: ConfigMap
+metadata:
+  name: local-volume-config
+  namespace: kube-system
+data:
+ storageClassMap: |
+    local-fast:
+      hostDir: "/mnt/ssds"
+    local-slow:
+      hostDir: "/mnt/hdds"
+    local_dynamic:
+      storageSource:
+        - "/dev/sdb"
+        - "/dev/sdc"
+```
+For LVM, its items are disk partitions, and for other mechanisms, they could be other things.
+StorageManager should know the scope and what to do with the input.
+
+Major downsides of this approach:
+
+* As provisioners running on all nodes shall a same configuration ConfigMap,
+storage sources of their nodes are required to be configured in the same way.
+Take LVM as an example, for configmap above, disks used for provisioning are required
+to be partitioned as "/dev/sdb" and "/dev/sdc".
+
+##### Change in populator
+
+In addition to PV, the `populator` will also be extended to watch PVC objects of all namespaces.
+When it finds that a PVC is newly annotated with `annScheduledNode`, and its node name match the value
```

Comment:
It also needs to watch for the provisioner name to match itself.

Reply:
Oh, you are right, at the first beginning, I thought the annotation of provisioner is set by the provisioner itself, and forgot to correct it here later when I realize the fact that it's set by the controller.
The `permissions` thing below is for the same reason